Infrastructure updates - get_local_path, browse & delete in Spaces, etc

Hwesta opened the pull request:
Several infrastructure and miscellaneous updates in one, coming from AIP reingest work but independent of it.
- `browse` and `delete_path` are now implemented in protocol spaces, instead of in Space or Package, as they are very dependent on the protocol and should be handled by the appropriate space.
- `full_path` and `full_pointer_file_path` are now `@property`'s, since that's effectively  what they are
- Added `get_local_path` to help work with packages that are not store locally, but need to be worked with locally.  As the name implies, gets a local path for that package - either the actual local path, or by copying it to the SS internal location
- Pipelines store some pipeline API information so they can call pipeline APIs
- `compress_package` accepts an algorithm parameter

mcantelon committed:
Added API-related fields to pipeline model/form.

Add hostname, API key & username to Pipeline model & form. Display in
pipeline template, with default values if none provided. Allow pipelines
to post API information, but do not display it if queried. Try to guess
a pipeline's IP if it requests default setup.  Update migration.

Hwesta committed:
Fixed pointer file path

Fixed pointer file moving to be relative to the Location, not an absolute
path.

Hwesta committed:
Make some attribute functions into properties

Location.full_path, Package.full_path and Package.full_pointer_file_path
are effectively properties instead of functions, so convert them to behave
like properties. Add @property decorator to those functions, and update
where they're accessed.

Hwesta committed:
Add structure for handling local & remote packages

Package functions need to be able to work with a package locally (ie to
extract and return files), but the package may not be stored locally.  To
help solve this, add a Package.local_path temporary attribute.  Whenever a
function needs to work with a package locally, call Package.get_local_path,
which will populate and returne Package.local_path, fetch the package into
the SS internal location if necessary.  If the package is fetched,
local_path_location set.

Related changes:
* download_file endpoint update to check is_compressed
* download_file & LOM update for extract_file change
* many places that used full_path use get_local_path()

Hwesta committed:
Update browse to use child spaces

Browse functionality is dependent on what protocol the Space is, so
move the implementation to the child spaces.  If no child space
implementation is found, try browsing locally.

For PipelineFS browse, update to use ls and parse the output, instead of
running two python calls.

Hwesta committed:
Minor fixes

Hwesta committed:
compress_package accepts compression algorithm parameter

Add compression constants to Package class for algorithms supported by
Archivematica.  Update compress_package to take a parameter for which
compression algorithm to run and use that.

Hwesta committed:
Update delete to use child spaces

Delete functionality is dependent on what protocol the Space is, so
move the implementation to the child spaces.  If no child space
implementation is found, try deleting the path locally.

Hwesta committed:
Delete staging copy when moving from storage service

mistydemeo commented on storage_service/locations/models.py:
```diff
@@ -1283,19 +1404,19 @@ def store_aip(self, origin_location, origin_path):
         # internal_usage_location/uuid/split/into/chunks/pointer.xml
```
Update comment too?

Hwesta replied:
Done

mistydemeo commented on storage_service/locations/models.py:
```diff
@@ -1164,26 +1231,45 @@ class Package(models.Model):
-            os.path.join(self.current_location.full_path(), self.current_path))
+            os.path.join(self.current_location.full_path, self.current_path))
 
+    @property
```
Move 1b03c073cdf6676c9f9df1f87cbf822d06455c3e after this commit so it can take advantage of this?

mistydemeo commented on storage_service/locations/api/resources.py:
```diff
@@ -122,10 +120,20 @@ class Meta:
+        # Try to guess Pipeline's IP, if doing default setup
+        if create_default_locations:
+            ip = bundle.request.META['REMOTE_ADDR'] or None
+            bundle.obj.remote_name = ip
```
It's okay for this to be `None`?

Hwesta replied:
No, but this is a best guess attempt.  `Pipeline.remote_name` defaults to None, and remote calls will not work if it's unset.  However, it can be set by the user - this is just a convenience to set it automatically if possible.

mistydemeo replied:
Ah, okay.

mistydemeo commented on storage_service/locations/api/resources.py:
```diff
@@ -122,10 +120,20 @@ class Meta:
         create_default_locations = bundle.data.get('create_default_locations', False)
+        # Try to guess Pipeline's IP, if doing default setup
+        if create_default_locations:
+            ip = bundle.request.META['REMOTE_ADDR'] or None
```
`bundle.request` always has this key?

Hwesta replied:
Probably, but using `.get` is safer.  Fixing.

Hwesta replied:
Done

mistydemeo commented on storage_service/locations/models.py:
```diff
@@ -1224,6 +1311,40 @@ def get_download_path(self, lockss_au_number=None):
             path = full_path
         return path
 
+    def get_local_path(self):
```
Maybe split this into a couple functions? This can cause a potentially very slow operation, which isn't necessarily what I'd expect from something called `get_local_path` (which implies it just returns a path, rather than copies).

e.g. maybe a function to check if there is a local path, and a version of this that returns a local path w/o copying.

Hwesta replied:
Hmm. I realize calling a big, slow copy operation is bad, but part of my goal with get_local_path was to abstract as much as possible from the calling functions, so they din't have to worry about checking for a local path and then getting it.

When would you expect there to be a local path when the package isn't stored locally (or locally mounted)?  How would we get a local path without copying?

An earlier version of this checked for a copy of this package in SS_INT and used that if found, but that was problematic since, for example, check_fixity would extract the package, reingest would find the local copy and start to use it, check_fixity would delete local copy when done, and reingest would break. Hence why this creates a new temp dir every time.

(Another problem I ran into was fetching a single file from an uncompressed AIP stored remotely triggered a full package copy, because is_compressed needs to run os.path.isdir locally, but that's a different problem.)

Thoughts? Suggestions?  Maybe rename this to `get_or_fetch_local_path`?  I'm unsure the best way to solve this - this was just my first attempt. :)

mistydemeo replied:
> part of my goal with get_local_path was to abstract as much as possible from the calling functions, so they din't have to worry about checking for a local path and then getting it.

My interpretation of the name `get_local_path` was initially that it was simply a getter - like `get_full_path` and friends. I didn't realize it was copying files until I read the code. So, my #1 suggestion would be that this be renamed to something like `copy_to_local_path`so it's clear.

Since there are a couple of code paths that set an attribute and then return a path without expensive disk operations, I think it'd also make sense to expose that distinctly from the rest of this. I could see us making the decision at some point to decide to perform an operation if there's a local path, but skip it entirely otherwise in order to avoid the file copy. Could either do that by making that `get_local_path`, which `copy_to_local_path` checks first as a convenience, or by having 100% separate functions.

Hwesta replied:
I like that idea - fetch most of the time, but expose the ability to decide not to if it's expensive.

I'm hesitant to call it `copy_to_local_path` since it may not return a copy (if it's found locally) and I don't want originals getting deleted accidentally. If you want a guaranteed copy, `extract_file` does that (somewhat oddly, now that I think about it. If local and uncompressed will copy it for sure, otherwise returns uncompressed copy). Though, extract_file calls get_local_path...

Maybe `try_local_path` which returns local path or None (or raises an exception?), and `fetch_local_path` which always returns a local path (calls check_local_path, may be the original copy).

mistydemeo replied:
How about `get_local_path` (which gets a local path attribute if there is one, and hence makes sense to me as a `get_` method) and `fetch_local_path`?

mistydemeo commented on storage_service/locations/models.py:
```diff
@@ -1507,7 +1669,7 @@ def check_fixity(self):
             message = failure.message
 
-        if self.is_compressed():
+        if delete_after and (self.local_path_location != self.current_location or self.local_path != self.full_path):
```
What's the reason for the secondary condition here?

Hwesta replied:
It should only delete the package if we know it's a copy (created from get_local_path or uncompression or something). This is to prevent deleting an uncompresed package stored locally.

mistydemeo commented on storage_service/locations/models.py:
```diff
@@ -1393,14 +1518,18 @@ def extract_file(self, relative_path='', extract_path=None):
         return (output_path, extract_path)
 
-    def compress_package(self, extract_path=None):
+    def compress_package(self, algorithm, extract_path=None):
```
Make `algorithm` a kwarg with default of `Package.COMPRESSION_TAR`?

Hwesta replied:
Why tar as the default? It seems to me if you're compressing something you'd want to specify how you're compressing it.  (And tar isn't technically compression. ;) )

mistydemeo replied:
Because then you don't have to change the places that are calling it. ;)

Hwesta replied:
Haha, fair, but when reading the code I'd rather know what compression algorithm is being used explicitly, rather than going look for the default param.

mistydemeo commented on storage_service/locations/models.py:
```diff
@@ -1393,14 +1518,18 @@ def extract_file(self, relative_path='', extract_path=None):
 
         If `extract_path` is provided, will compress there, otherwise to a temp
-        directory.
+        directory in the SS internal location.
```
Update docstring with `algorithm` too

Hwesta replied:
Updated this to use Sphinx-style docstring.

mistydemeo commented:
:thumbsup:

Hwesta committed:
Package: Split get_local_path to get/fetch_local_path

get_local_path does not imply it may do a copy. Split it into two
functions, get_local_path which returns a local path only if it already
exists (or None), and fetch_local_path with the old behaviour of
copying the package locally if it isn't found. Update references to use
fetch_local_path.

Hwesta committed:
Space: Move path mangling to Space

move_to_storage_service expects the source path to be absolute, and the
destination path to be absolute inside the staging path. Since this is
common, move it to Space.

mistydemeo commented:
Updates look good! Don't have any more comments besides the ones I already had.
